opentracing-util/src/main/java/io/opentracing/util/ThreadLocalScope.java

```diff
@@ -48,7 +49,13 @@ public void close() {
             wrapped.finish();
         }
 
-        scopeManager.tlsScope.set(toRestore);
+        if (toRestore != null) {
+            scopeManager.tlsScope.set(toRestore);
+            scopeManager.listener.onActivated(toRestore.wrapped);
+        } else {
+            scopeManager.tlsScope.remove();
+            scopeManager.listener.onClosed();
```

Comment:
the previous comments got lost - the onClosed must be called regardless of the previous span presence.

Reply:
And thats exactly what i don't want to do, see discussion in #334.
Most common use-case is setting MDC (or ThreadContext for log4j) - and that uses copy-on-write immutable maps.
So, having to do following sequence:
1) onActivated(1) - set MDC (copy internal map)
2) onClose(1) - delete MDC (copy internal map) - **useless**
3) onActivated(2) - set MDC (copy internal map)
4) onClose(2) - delete MDC (copy internal map) - **useless**
5) onActivated(3) - set MDC (copy internal map)
wastes resources. Consumer cannot know, whether onActivated is coming right after onClose.
Thats why originally wanted only protected method for simple extension, or single method interface, so consumer can decide, what is best. This API should be simplistic, if you need something complex, you can copy whole class.

Reply:
BTW thats why I originally proposed single method interface, with possible null as argument - its ugly, but efficient and versatile.

Reply:
You're making assumption that MDC is the only use case. If you don't want to clear MDC from onClosed(), you don't have to, it does not negate the usefulness of the callback for other use cases. For example, see https://github.com/census-instrumentation/opencensus-specs/issues/247

Reply:
You have to clear it in onClosed(), because you can't leave it set on last deactivation (updated example above, sorry I left it out).
Not saying its only use case, but its use case that should be supported. Give users freedom to do what they need, don't force them to follow exact and strict pattern.

Only model, that would actually fulfill both use-cases IMO, is ugly but flexible:
```
interface ScopeListener {
  void onSpanChanged(@Nullable Span activeSpan, @Nullable Span deactivatedSpan);
}
```

Or, I would revisit protected method and simply inheritance - do everyone whatever they want.
```
class ThreadLocalScopeManager {
  protected void setThreadScope(ThreadLocalScope scope) {
    this.tlsScope.set(scope);
  }
}
```
this can serve as "around advice". And it would make the class extensible in many ways, which I don't think is a bad thing.

Reply:
Moved discussion to bottom, it will get lost here on another push.